utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+     * <p>The returned future will be completed exceptionally if the downstream subscriber cancels the subscription, or
+     * if the {@code complete} call was performed after a {@code complete} or {@link #error(Throwable)} call.
+     *
+     * @return A future that is completed when the complete has been sent to the downstream subscriber.
+     */
+    public CompletableFuture<Void> complete() {
+        log.trace(() -> "Received complete()");
+
+        OnCompleteQueueEntry<T> entry = new OnCompleteQueueEntry<>();
+
+        try {
+            checkRejectException();
+            setRejectExceptionOrThrow(() -> new IllegalStateException("complete() has been invoked"));
+            eventQueue.add(entry);
+            processEventQueue();
+        } catch (RuntimeException t) {
+            entry.resultFuture.completeExceptionally(t);
+        }
+        return entry.resultFuture;
+    }
+
+    /**
+     * Indicate that no more {@link #write(Object)} calls will be made, and that streaming of messages has failed.
+     *
+     * <p>This can be called before any in-flight {@code write} calls are complete. Such messages will be processed before the
+     * stream is treated as being in-error. The returned future will be completed successfully when the {@code error} is
+     * sent to the downstream subscriber.
+     *
+     * <p>After this method is invoked, any future {@link #write(Object)}, {@link #complete()} or {@code #error(Throwable)}
+     * calls will be completed exceptionally and not be processed.
+     *
+     * <p>The returned future will be completed exceptionally if the downstream subscriber cancels the subscription, or
+     * if the {@code complete} call was performed after a {@link #complete()} or {@code error} call.
+     *
+     * @param error The error to send.
+     * @return A future that is completed when the exception has been sent to the downstream subscriber.
+     */
+    public CompletableFuture<Void> error(Throwable error) {
+        log.trace(() -> "Received error() with " + error, error);
+
+        OnErrorQueueEntry<T> entry = new OnErrorQueueEntry<>(error);
+
+        try {
+            checkRejectException();
+            setRejectExceptionOrThrow(() -> new IllegalStateException("error() has been invoked"));
+            eventQueue.add(entry);
+            processEventQueue();
+        } catch (RuntimeException t) {
+            entry.resultFuture.completeExceptionally(t);
+        }
+        return entry.resultFuture;
+    }
+
+    /**
+     * A method called by the downstream subscriber in order to subscribe to the publisher.
+     */
+    @Override
+    public void subscribe(Subscriber<? super T> s) {
+        if (subscriber != null) {
+            s.onSubscribe(new NoOpSubscription());
+            s.onError(new IllegalStateException("Only one subscription may be active at a time."));
+        }
+        this.subscriber = s;
+        s.onSubscribe(new SubscriptionImpl());
+        processEventQueue();
```

Comment:
Do we need to call `processEventQueue` here?

Reply:
Yes. We support writing to this publisher before anyone subscribes, and we want any complete() or error() calls to be sent as soon as we get a subscriber.